core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1682,7 +1682,10 @@ private <C extends Comparable<C>> RexNode simplifyAnd2ForUnknownAsFalse(
             final RexLiteral literal = comparison.literal;
             final RexLiteral prevLiteral =
                 equalityConstantTerms.put(comparison.ref, literal);
-            if (prevLiteral != null && !literal.equals(prevLiteral)) {
+
+            if (prevLiteral != null
+                && literal.getType().equals(prevLiteral.getType())
+                && !literal.equals(prevLiteral)) {
```

Comment:
so two literals can be equal but their types can be different?
one could argue that this is a bug in SqlLiteral::equals.

Reply:
the situation was caused by two literals which had the same value; but different types (`VARCHAR`; `CHAR(n)`)

actually the predicate based simplification also does this (beyond other things); but there are situations in which its disabled.

Reply:
I think `VARCHAR` and `CHAR` should be considered equivalent if RexLiteral.value is the same

Reply:
CHAR(N) indicates that the literal has to be padded with spaces, whereas VARCHAR indicates that trailing spaces are to be ignored. So two literals with identical strings but different types can have different values.